application/features/bootstrap/EmailContext.php

```diff
@@ -566,19 +572,30 @@ public function aBackupCodeMfaOptionDoesNotExist()
     /**
      * @Given there are :arg1 backup codes
      */
-    public function thereAreXBackupCodes($count)
+    public function thereAreXBackupCodes($desiredCount)
     {
         $backupMfa = $this->getMfa(Mfa::TYPE_BACKUPCODE);
 
         if (empty($backupMfa)) {
-            if ($count == 0) {
+            if ($desiredCount == 0) {
                 return;
             }
-            throw new InvalidArgumentException('There is no MFA Backup Code option to be able to create ' . $count . ' codes for.');
+            throw new InvalidArgumentException('There is no MFA Backup Code option available for the test.');
         }
 
-        MfaBackupcode::createBackupCodes($backupMfa->id, $count);
-        $backupMfa->refresh();
+        $codeCount = count($this->tempBackupCodes);
+
+        Assert::greaterThanEq($codeCount, $desiredCount, 'There are not enough backup codes to run the test.');
+
+        $countDiff = $codeCount - $desiredCount;
+
+        for ($i = 0; $i < $countDiff; $i++) {
+            $nextCode = array_shift($this->tempBackupCodes);
+            MfaBackupcode::validateAndRemove($backupMfa->id, $nextCode, False);
```

Comment:
You may be able to simplify some of this code by simply blowing away any existing backup codes and re-creating exactly how many you need. That might take less code that looping through and calling `validateAndRemove(...)` for each backupcode that you don't need. Just a suggestion.

Reply:
Never mind. I played around with this idea a little bit more, and since you need to both (A) have an exact number of backup codes remaining and (B) know what they are, it looks like the way you're currently doing it is better than what I suggested.